File: Python/115__Distinct_Subsequences.py
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        m, n = len(s), len(t)

        if m < n:
            return 0

        dp = [[0] * (m + 1) for _ in range(n + 1)]
        dp[0] = [1] * (m + 1)

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if s[j - 1] == t[i - 1]:
                    dp[i][j] = dp[i][j - 1] + dp[i - 1][j - 1]
                else:
                    dp[i][j] = dp[i][j - 1]
        return dp[-1][-1]

File: Java/115__Distinct_Subsequences.java
class Solution {
    public int numDistinct(String s, String t) {
        int dp[][] = new int[s.length() + 1][t.length() + 1];
        for (int i = 0; i <= s.length(); i++)
            dp[i][0] = 1;
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < t.length(); j++) {
                dp[i + 1][j + 1] = dp[i][j + 1];
                if (s.charAt(i) == t.charAt(j))
                    dp[i + 1][j + 1] += dp[i][j];
            }
        }
        return dp[s.length()][t.length()];
    }
}
